File: examples/high_quality_solution/espressif/components/esp_board_manager/devices/dev_gpio_ctrl/dev_gpio_ctrl.py
# SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
# SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
#
# See LICENSE file for details.

# GPIO control device config parser
VERSION = 'v1.0.0'

import sys

def get_includes():
    """Get required header includes for GPIO control device"""
    return ['dev_gpio_ctrl.h']

def parse(name, config, peripherals_dict=None):
    """
    Parse GPIO control device configuration from YAML to C structure

    Args:
        name (str): Device name
        config (dict): Device configuration dictionary
        peripherals_dict (dict): Dictionary of available peripherals for validation

    Returns:
        dict: Parsed configuration with 'struct_type' and 'struct_init' keys
    """
    # Extract configuration parameters
    device_config = config.get('config', {})
    peripherals_list = config.get('peripherals', [])

    # GPIO control configuration
    active_level = device_config.get('active_level', 1)
    default_level = device_config.get('default_level', 0)

    # Extract GPIO peripheral name from peripherals list
    gpio_name = 'gpio-0'  # Default fallback
    if peripherals_list and len(peripherals_list) > 0:
        gpio_periph = peripherals_list[0]
        if isinstance(gpio_periph, dict) and 'name' in gpio_periph:
            gpio_periph_name = gpio_periph['name']
        else:
            gpio_periph_name = str(gpio_periph)

        if gpio_periph_name.startswith('gpio') or gpio_periph_name.startswith('gpio_'):
            # Check if peripheral exists in peripherals_dict if provided
            if peripherals_dict is not None and gpio_periph_name not in peripherals_dict:
                raise ValueError(f"GPIO device {name} references undefined peripheral '{gpio_periph_name}'")
            gpio_name = gpio_periph_name
        else:
            raise ValueError(f'GPIO device {name} should reference a GPIO peripheral, got: {gpio_periph_name}')
    else:
        raise ValueError(f'GPIO device {name} must have at least one GPIO peripheral defined')

    # Build structure initialization
    struct_init = {
        'name': f'"{name}"',
        'type': str(config.get('type', 'gpio_ctrl')),
        'gpio_name': gpio_name,
        'active_level': active_level,
        'default_level': default_level
    }

    return {
        'struct_type': 'dev_gpio_ctrl_config_t',
        'struct_init': struct_init
    }

File: examples/high_quality_solution/espressif/components/esp_board_manager/devices/dev_gpio_ctrl/dev_gpio_ctrl.c
/*
 * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO., LTD
 * SPDX-License-Identifier: LicenseRef-Espressif-Modified-MIT
 *
 * See LICENSE file for details.
 */

#include "esp_log.h"
#include "esp_err.h"
#include "dev_gpio_ctrl.h"
#include "driver/gpio.h"
#include "esp_board_periph.h"
#include "esp_board_device.h"
#include <string.h>

static const char *TAG = "DEV_GPIO_CTRL";

int dev_gpio_ctrl_init(void *cfg, int cfg_size, void **device_handle)
{
    if (!cfg || !device_handle) {
        ESP_LOGE(TAG, "Invalid parameters, cfg: %p, device_handle: %p", cfg, device_handle);
        return -1;
    }

    const dev_gpio_ctrl_config_t *config = (const dev_gpio_ctrl_config_t *)cfg;

    // Get GPIO peripheral handle
    periph_gpio_handle_t *gpio_handle = NULL;
    if (!config->gpio_name || strlen(config->gpio_name) == 0) {
        ESP_LOGE(TAG, "No GPIO peripheral name configured for device: %s", config->name);
        return -1;
    }

    int ret = esp_board_periph_ref_handle(config->gpio_name, (void*)&gpio_handle);
    if (ret != 0) {
        ESP_LOGE(TAG, "Failed to get GPIO peripheral handle '%s': %d", config->gpio_name, ret);
        return -1;
    }

    // Get GPIO peripheral config
    periph_gpio_config_t gpio_config = {0};
    esp_err_t config_ret = esp_board_periph_get_config(config->gpio_name, (void*)&gpio_config);
    if (config_ret != ESP_OK) {
        ESP_LOGE(TAG, "Failed to get GPIO peripheral config '%s': %s", config->gpio_name, esp_err_to_name(config_ret));
        return -1;
    }

    // Set the active level using the GPIO handle
    esp_err_t err = gpio_set_level(gpio_handle->gpio_num, config->active_level);
    if (err != ESP_OK) {
        ESP_LOGE(TAG, "Failed to set active level for GPIO %d: %s", gpio_handle->gpio_num, esp_err_to_name(err));
        return err;
    }

    ESP_LOGI(TAG, "Initialized GPIO control device '%s': gpio=%d, set initial level:%d (active_level=%d, default_level=%d)",
             config->name, gpio_handle->gpio_num, config->active_level, config->active_level, config->default_level);

    // Store the GPIO handle for later use
    *device_handle = gpio_handle;
    return 0;
}

int dev_gpio_ctrl_deinit(void *device_handle)
{
    if (!device_handle) {
        ESP_LOGE(TAG, "Invalid handle");
        return -1;
    }
    const char *name = NULL;
    const esp_board_device_handle_t *device_handle_struct = esp_board_device_find_by_handle(device_handle);
    if (device_handle_struct) {
        name = device_handle_struct->name;
    }
    dev_gpio_ctrl_config_t *cfg = NULL;
    esp_board_device_get_config(name, (void **)&cfg);
    if (cfg) {
        esp_board_periph_unref_handle(cfg->gpio_name);
    }

    // Cast back to GPIO handle type
    periph_gpio_handle_t *gpio_handle = (periph_gpio_handle_t *)device_handle;
    if (!gpio_handle) {
        ESP_LOGE(TAG, "Invalid GPIO handle");
        return -1;
    }
    ESP_LOGI(TAG, "GPIO control device gpio %d deinitialized", gpio_handle->gpio_num);
    return 0;
}
